### PR TITLE
Validate HuggingFace API filenames before using them as local paths

### PR DESCRIPTION
**Pull request by michaelneale**
Minor nit-fix - just to protect paths. 

This finding was discovered by project loupe.

### CODE REVIEW

File: crates/goose/src/providers/local_inference/hf_models.rs [hunk 1]
```diff
@@ -375,6 +375,26 @@ fn build_download_url(repo_id: &str, filename: &str) -> String {
...
+    Path::new(&normalized)
+        .components()
+        .all(|component| matches!(component, Component::Normal(_) | Component::CurDir))
+}
+
+fn sanitize_siblings(siblings: Vec<HfApiSibling>) -> Vec<HfApiSibling> {
+    siblings
+        .into_iter()
+        .filter(|s| is_safe_relative_path(&s.rfilename))
+        .collect()
```
[the remark this answers is not in the document]
Comment (DOsinga) [reply crates/goose/src/providers/local_inference/hf_models.rs:412]: Good catch — addressed in 952efd5787c09122ba07794e2f483e18140cb256. `should_download_for_mlx` now rejects filenames that fail `is_safe_relative_path`, so `mlx_download_filenames` filters out traversal/absolute entries before `resolve_mlx_model` downloads them. Added a unit test covering this.

File: crates/goose/src/providers/local_inference/hf_models.rs [hunk 2]
```diff
@@ -1766,6 +1816,9 @@ fn merge_model_info(existing: &mut HfModelInfo, duplicate: HfModelInfo) {
 }
 
 fn should_download_for_mlx(filename: &str) -> bool {
+    if !is_safe_relative_path(filename) {
+        return false;
```
[the remark this answers is not in the document]
Comment (DOsinga) [reply crates/goose/src/providers/local_inference/hf_models.rs]: Fixed in c6283aa22. Siblings are now sanitized at the entry points (`mlx_variants_from_model_info` and `resolve_mlx_model`) before compatibility detection, so a repo whose required MLX weights only exist under unsafe names no longer passes `is_mlx_compatible_repo` — it yields no variant and resolution fails cleanly instead of downloading a broken snapshot. Test updated accordingly.

File: crates/goose/src/providers/local_inference/hf_models.rs [hunk 3]
```diff
@@ -686,7 +714,7 @@ pub async fn get_repo_gguf_variants(repo_id: &str) -> Result<Vec<HfQuantVariant>
     }
 
     let model: HfApiModel = response.json().await?;
-    let siblings = model.siblings.unwrap_or_default();
+    let siblings = sanitize_siblings(model.siblings.unwrap_or_default());
```
[the remark this answers is not in the document]
Comment (DOsinga) [reply crates/goose/src/providers/local_inference/hf_models.rs:737]: Fixed in 3fe903921. `group_into_variants` now skips shard groups that are not a complete contiguous `-of-N` set (new `shards_are_complete` helper), so the listing/search APIs no longer advertise a sharded quant whose shards were dropped by sanitization (or were simply missing upstream). The download-side completeness check in `resolve_model_spec_full` stays as the authoritative gate with its detailed diagnostics. Added a test covering the incomplete-shard case.

File: crates/goose/src/providers/local_inference/hf_models.rs [hunk 4]
```diff
@@ -375,6 +375,49 @@ fn build_download_url(repo_id: &str, filename: &str) -> String {
...
+    use std::path::{Component, Path};
+
+    if filename.is_empty() {
+        return false;
+    }
+
+    let normalized = filename.replace('\\', "/");
+    Path::new(&normalized)
+        .components()
+        .all(|component| matches!(component, Component::Normal(_) | Component::CurDir))
```
[the remark this answers is not in the document]
Comment (DOsinga) [reply crates/goose/src/providers/local_inference/hf_models.rs]: Fixed in 611d9d13a. `is_safe_relative_path` now splits on `/` (the same way `snapshot_root_for_file` counts segments) and rejects empty, `.`, and `..` segments — so `./model.safetensors` and `weights/./model.safetensors` are no longer treated as safe, keeping validation consistent with the pop logic. Tests updated.

**Comment by DOsinga** [comment]
Ok, I'm calling it on this one. codex keeps finding corner cases here, this is not the trivial fix we thought it was. let's go back to the drawing board and figure out something more stable
